dialectic-reconnect/src/retry/end.rs

```diff
@@ -0,0 +1,313 @@
+    time::Instant,
+};
+
+use crate::{
+    retry::{ConnectTo, Init, ReconnectStrategy, Retry, RetryError, RetryStrategy},
+    util::{sleep_until_or_deadline, timeout_at_option},
+};
+
+/// The insides of the [`Sender`] and [`Receiver`] are almost identical, and the functionality we
+/// need is almost the same. The `End` struct is the insides of both `Sender` and `Receiver`: all
+/// that  differs is the choice of type parameter.
+#[Transmitter(Tx)]
+#[Receiver(Rx)]
+pub struct End<H: Session, Address, Key, Err, ConnectErr, HandshakeErr, Inner, Other, Tx, Rx> {
+    /// The address to connect to
+    address: Arc<Address>,
+    /// The key for this session, if one has been assigned.
+    key: Arc<Mutex<Option<Key>>>,
+    /// The current inner thing (a `tx` or `rx` connection end).
+    inner: Option<Inner>,
+    /// A queue of next inner things, from which we will pop when necessary.
+    next_inner: Arc<Mutex<VecDeque<Inner>>>,
+    /// A queue of next other things, to which we will push when we create a new connection.
+    next_other: Weak<Mutex<VecDeque<Other>>>,
+    /// An async function to create a new connection, or return an error.
+    connect: Arc<ConnectTo<Address, ConnectErr, Tx, Rx>>,
+    /// An async function to perform an initial handshake.
+    init: Arc<Init<H, Key, HandshakeErr, Tx, Rx>>,
+    /// An async function to perform an initial handshake.
+    retry: Arc<Retry<H, Key, HandshakeErr, Tx, Rx>>,
+    /// A function describing the desired retry strategy for errors in the underlying connection.
+    recover: Arc<dyn Fn(usize, &Err) -> RetryStrategy + Sync + Send>,
+    /// A function describing the desired retry strategy for errors while attempting to establish a
+    /// connection.
+    recover_connect: Arc<dyn Fn(usize, &ConnectErr) -> ReconnectStrategy + Sync + Send>,
+    /// A function describing the desired retry strategy for errors while attempting to perform a
+    /// handshake.
+    recover_handshake: Arc<dyn Fn(usize, &HandshakeErr) -> ReconnectStrategy + Sync + Send>,
+    /// An optional timeout, which bounds all retry attempts.
+    timeout: Option<Duration>,
+}
+
+pub type SenderEnd<H, Address, Key, ConnectErr, HandshakeErr, Tx, Rx> =
+    End<H, Address, Key, <Tx as Transmitter>::Error, ConnectErr, HandshakeErr, Tx, Rx, Tx, Rx>;
+
+pub type ReceiverEnd<H, Address, Key, ConnectErr, HandshakeErr, Tx, Rx> =
+    End<H, Address, Key, <Rx as Receiver>::Error, ConnectErr, HandshakeErr, Rx, Tx, Tx, Rx>;
+
+#[Transmitter(Tx)]
+#[Receiver(Rx)]
+pub struct Recoveries<ConnectErr, HandshakeErr, Tx, Rx> {
+    pub connect: Arc<dyn Fn(usize, &ConnectErr) -> ReconnectStrategy + Sync + Send>,
+    pub handshake: Arc<dyn Fn(usize, &HandshakeErr) -> ReconnectStrategy + Sync + Send>,
+    pub tx: Arc<dyn Fn(usize, &Tx::Error) -> RetryStrategy + Sync + Send>,
+    pub rx: Arc<dyn Fn(usize, &Rx::Error) -> RetryStrategy + Sync + Send>,
+}
+
+#[Transmitter(Tx)]
+#[Receiver(Rx)]
+#[allow(clippy::type_complexity)]
```

Comment:
:sweat_smile: :sweat_smile: :sweat_smile:

Reply:
Hahaha, yeah, the type is not ideal here. Wish there was a way to reduce the number of parameters in play, but I'm not sure there is...